### PR TITLE
[spirv] support enum type

### LINKED ISSUE #2317: [SPIRV] Type enum is unimplemented
Very interesting issue! 
While compiling HLSL to SPIRV shader, The DXC fails to compile enum types, though it compiles fine for DXIL output format. Do you think this is a quick fix and, we can expect in coming weeks? or it's gonna take long time to fix it?
Here is the Permalink : http://shader-playground.timjones.io/0cf75398d6e0928e2494ee92f432781c

### PR DESCRIPTION
**Pull request by jaebaek**
Fixes #2317

### CODE REVIEW

File: tools/clang/test/CodeGenSPIRV/type.enum.hlsl [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+      d = First;
+      break;
+    default:
+      d = Third;
+      break;
+  }
+
+//CHECK:      [[fourth:%\d+]] = OpLoad %int %Fourth
+//CHECK-NEXT:                   OpStore %e [[fourth]]
+  static ::Number e = Fourth;
```
Comment (ehsannas) [inline tools/clang/test/CodeGenSPIRV/type.enum.hlsl]: Why are you using `static` everywhere?

File: tools/clang/test/CodeGenSPIRV/type.enum.hlsl [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  int foo = a;
+
+//CHECK:      [[fourth:%\d+]] = OpLoad %int %Fourth
+//CHECK-NEXT:                   OpStore %b [[fourth]]
+  b = Fourth;
+
+//CHECK:          [[c:%\d+]] = OpAccessChain %_ptr_Uniform_int %c %uint_0
+//CHECK-NEXT: [[third:%\d+]] = OpLoad %int %Third
+//CHECK-NEXT:                  OpStore [[c]] [[third]]
+  c.Append(Third);
```
Comment (ehsannas) [inline tools/clang/test/CodeGenSPIRV/type.enum.hlsl]: Can you try `Number::Third` instead of `Third` ?

**Comment by sunillnaik** [comment]
I think I can pull the fix now! though not merged to master yet. ;)

**Comment by jaebaek** [comment]
@sunillnaik sorry for being late to handle this. I had some urgent stuff. I am working on it now.

File: tools/clang/lib/SPIRV/SpirvEmitter.cpp [hunk 3]
```diff
@@ -5012,7 +5019,8 @@ void SpirvEmitter::storeValue(SpirvInstruction *lhsPtr,
   const bool lhsIsFloatMat = lhsIsMat && matElemType->isFloatingType();
   const bool lhsIsNonFpMat = lhsIsMat && !matElemType->isFloatingType();
 
-  if (isScalarType(lhsValType) || isVectorType(lhsValType) || lhsIsFloatMat) {
+  if (isScalarType(lhsValType) || isVectorType(lhsValType) || lhsIsFloatMat ||
+      isEnumType(lhsValType)) {
```
[the remark this answers is not in the document]
Comment (jaebaek) [reply tools/clang/lib/SPIRV/SpirvEmitter.cpp]: It makes sense. Updated.

File: tools/clang/test/CodeGenSPIRV/type.enum.hlsl (hunk 2 above)
Comment (jaebaek) [reply tools/clang/test/CodeGenSPIRV/type.enum.hlsl]: done

File: tools/clang/test/CodeGenSPIRV/type.enum.hlsl (hunk 1 above)
Comment (jaebaek) [reply tools/clang/test/CodeGenSPIRV/type.enum.hlsl]: I tested
```
static ::Number a = Second;
groupshared Number b;
AppendStructuredBuffer<Number> c;
```

and in `main()`,
```
  Number d;
  static ::Number e = Fourth;
```

I just want to check many variable scope.
Is there a specific scope (or keyword) you want to test?

**Comment by sunillnaik** [comment]
I pulled commit 0437e0b096667335a36a93dc9faae4104f49f1fe and it works for me, thanks :)
👍

**Pull request merged**